[Zheqi] ip

adamzzq opened the pull request.

adamzzq committed:
Level-0: Rename to Adam, add greeting and exit

adamzzq committed:
Add echo and exit feature

adamzzq committed:
Implement add and list feature

Add the functionality to store user inputs and display back to the user
on command in the form of an indexed list.

adamzzq committed:
Add mark and unmark feature

User can mark a task as done or revert back the status to not done.
The corresponding task and its status are displayed as an echoed
message. The status of all tasks are shown in the list on command.

adamzzq committed:
Improve code quality

Enhance code readability through keyword changes, variable renaming and
redundancy removal, in order to follow the coding standard.

adamzzq committed:
Customize chatbot prompts

adamzzq committed:
Bug fix: Prevent incorrect task status changes

Ensure that mark/unmark command does not erroneously revert the status
of a task.

adamzzq committed:
Add ToDo, Deadline, Event classes

adamzzq committed:
Implement Todo, Event, Deadline functionalities

User can add tasks with starting and ending time as events, with only
ending time as deadlines, and without starting/ending time as todos

adamzzq committed:
Add minimal exception handling in Adam

Wrap the returned object of error-prone methods in Optional<>. This is
to deal with the removal of the add task feature previously implemented
which accepts any inputs as tasks.

adamzzq committed:
Enhance: Customize messages

Inject humor into response messages for list and add task commands.

adamzzq committed:
Refactor: Rename variables to enhance code readability

adamzzq committed:
Reconstruct Parser as enum and restructure related classes

adamzzq committed:
Remove redundant import statements and unused variables

adamzzq committed:
Configure text-ui testing

* rename Duke to Adam in the bash script
* updated input and expected output

dylansiew commented on src/main/java/Adam.java:
```diff
@@ -0,0 +1,25 @@
+            }
+
+            System.out.println(Message.DELIMITER);
+        }
```
Good use of constants, remember to close the scanner

dylansiew commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,3 @@
+public interface Command {
```
Good use of abstractions

dylansiew commented on src/main/java/TaskGenerator.java:
```diff
@@ -0,0 +1,10 @@
+    public static Task generate(String[] parsedInput) {
+        if (parsedInput.length == 1) {
+            return new ToDo(parsedInput[0]);
+        } else if (parsedInput.length == 2) {
```
Hardcoding to the length of the parsedInput could possibly cause issues in the future when adding more Tasks

dylansiew commented on src/main/java/Message.java:
```diff
@@ -0,0 +1,57 @@
+    public static final String UNMARK_MESSAGE = "Whoopsie-daisy! This task is back in action:\n";
+
+    public static final String ERROR_MESSAGE = "Oopsie! Looks like I got tangled up in my circuits.\n"
+            + "Could you please try again?";
```
Good implementation of constants

dylansiew reviewed (commented):
Overall very understandable code

joshualeejunyi commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.regex.Pattern;
+
+public enum Parser {
+    BYE("bye"),
+    LIST("list"),
+    TOGGLE("^(mark [0-9]*|unmark [0-9]*)"),
+    TODO("todo\\s(.+)"),
+    DEADLINE("deadline\\s(.+)/by\\s(.+)"),
+    EVENT("event\\s(.+)/from\\s(.+)/to\\s(.+)");
```
I like the use of enum to make it more readable!

joshualeejunyi commented on src/main/java/Adam.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Adam {
+    public static void main(String[] args) {
+        Scanner input = new Scanner(System.in);
```
Don't forget to close the scanner at the end so that you don't leak resources!

joshualeejunyi commented on src/main/java/CommandGenerator.java:
```diff
@@ -0,0 +1,19 @@
+
+public class CommandGenerator{
+    public static Optional<Command> generate(String input) {
+        switch (Parser.analyzeInput(input)) {
```
Just a nitpick, but might be better to split lines into more readable chunks. Noticed the same across the files as well :)
Something like: 
```suggestion
        Parser parsedInput = Parser.analyzeInput(input);
        switch (parsedInput) {
```

joshualeejunyi commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by + ")";
```
Minor nitpick, but why not String.format() if you have used it in Event.java? Would be great to standardize your code.

joshualeejunyi commented on src/main/java/Message.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Random;
+
+public class Message {
```
Really like this class of constants! I was considering something similar for my own iP as well

joshualeejunyi commented on src/main/java/TaskGenerator.java:
```diff
@@ -0,0 +1,10 @@
+public class TaskGenerator {
+    public static Task generate(String[] parsedInput) {
+        if (parsedInput.length == 1) {
+            return new ToDo(parsedInput[0]);
+        } else if (parsedInput.length == 2) {
+            return new Deadline(parsedInput[0], parsedInput[1]);
+        }
+        return new Event(parsedInput[0], parsedInput[1], parsedInput[2]);
+    }
```
Perhaps you can reuse enums that you have defined above to determine the type of task that you are generating?

joshualeejunyi commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public void displayAll() {
+        IntStream.rangeClosed(1, tasks.size())
+                .mapToObj(index -> index + "." + getTask(index))
+                .forEach(System.out::println);
```
I would suggest splitting these into multiple lines, or commenting on each line so that it will be easy to remember what it is actually doing!

joshualeejunyi reviewed (commented):
Overall, good effort! Some tweaking and cleaning up might be useful to make things easier to read and understand :)

adamzzq committed:
Refactor code according to reviews

* encapsulate scanner in CommandGenerator.java instead of Adam.java, so
it can be closed properly
* deprecate TaskGenerator.java; make generate() a method of TaskList
* use Parser enum to generate new Task to avoid comparing array length
which is not open to extension
* split long statements into multiple lines to improve readability

adamzzq committed:
Refactor: Correct input reading logic

* Restore Scanner to Adam.java from CommandGenerator.java, because it
is more logical to read input in the main method. CommandGenerator
should focus on its own job in creating new instances of commands
* Use exit flag to terminate loop and exit App, whereby scanner is
properly closed
* Command.execute() returns boolean to indicate whether to exit loop

yuhengr commented on src/main/java/Message.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Message {
+    public static final String DELIMITER = "\n>====> >====> >====> >====> >====> >====> >====> >====> >====>\n";
+
```
This is a great way to keep constants organised!

yuhengr commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,63 @@
+    TODO("todo\\s(.+)"),
+    DEADLINE("deadline\\s(.+)/by\\s(.+)"),
+    EVENT("event\\s(.+)/from\\s(.+)/to\\s(.+)");
+
```
Great work handling the inputs in an organised way

yuhengr commented on src/main/java/Message.java:
```diff
@@ -0,0 +1,58 @@
+            + "  \\ \\  \\ \\  \\ \\  \\_\\\\ \\ \\  \\ \\  \\ \\  \\    \\ \\  \\ \n"
+            + "   \\ \\__\\ \\__\\ \\_______\\ \\__\\ \\__\\ \\__\\    \\ \\__\\\n"
+            + "    \\|__|\\|__|\\|_______|\\|__|\\|__|\\|__|     \\|__|\n";
+
```
Perhaps the lineSeparator method could be used here? I believe it was mentioned that the behaviour of \n depends on the platform. Please correct me if I'm mistaken

yuhengr commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,63 @@
+        Matcher deadlineMatcher = deadlinePattern.matcher(input);
+        Matcher todoMatcher = todoPattern.matcher(input);
+        Matcher toggleMatcher = togglePattern.matcher(input);
+
```
Perhaps having some comments here for explanation will help as it seems like there is quite a bit of things happening

yuhengr reviewed (commented):
Overall great work in making use of OOP to abstract the different classes and ensuring that the code is well organised. Keep up the good work

adamzzq committed:
Refactor: Divide classes into packages

adamzzq committed:
Handle Error

Use AdamException to handle:
* invalid input
* index out of bound error in list and toggle command

adamzzq committed:
Merge branch 'branch-Level-5'

adamzzq committed:
Add help menu command

Notify the user to use help command at App start and when invalid
inputs are keyed in

adamzzq committed:
Merge branch 'branch-A-Packages'

adamzzq committed:
Improve UX

Allow more flexibility in commands format, the user can add any number
of white spaces between the task the time

adamzzq committed:
Add delete command and corresponding taskList method

adamzzq committed:
Implement delete task feature

adamzzq committed:
Update delete message and handle errors related to delete command

adamzzq committed:
Merge remote-tracking branch 'origin/branch-Level-6'

adamzzq committed:
Implement save and load feature

* Automatically load the task list from local data file on APP start.
* Automatically save to the local data file upon adding each task.
* Handle the case of no existing file by creating an empty data file.
* Display error message should the loading or saving fails.

adamzzq committed:
Edit help message

adamzzq committed:
Merge branch 'branch-Level-6'

adamzzq committed:
Merge branch 'branch-Level-7'

adamzzq committed:
Fix scanner bug

adamzzq committed:
Merge branch 'branch-Level-7'

adamzzq committed:
Update file related error message

adamzzq committed:
Refactor: Simply parser matching logic

adamzzq committed:
Refactor: Simply parser matching logic

adamzzq committed:
Merge remote-tracking branch 'origin/master'

Jonoans commented on src/main/java/ui/Message.java:
```diff
@@ -0,0 +1,98 @@
+
+import java.util.Random;
+
+public class Message {
```
Good job with the use of a constants class :D

Jonoans reviewed (commented):
Really well done! Not much issues. Good use of constants. Methods are short and easy to digest as individual units which helps one to better understand how code units work together to compose functionalities, good job!

adamzzq committed:
Refactor: Extract Ui, Storage, Parser, and TaskList classes for better
modularity and clarity.

adamzzq committed:
Add Time class for date and time parsing and formatting

adamzzq committed:
Add findTasks method to TaskList class

adamzzq committed:
Merge pull request #2 from adamzzq/branch-Level-8

Add Time class for date and time parsing and formatting

adamzzq committed:
Merge pull request #3 from adamzzq/branch-Level-9

Add findTasks method to TaskList class

adamzzq committed:
Add FindCommand and modify TaskList.findTasks() method

adamzzq committed:
Merge pull request #4 from adamzzq/branch-Level-9

Add FindCommand and modify TaskList.findTasks() method

adamzzq committed:
Add javadoc comments for classes and methods

adamzzq committed:
Merge pull request #5 from adamzzq/branch-A-JavaDoc

Add javadoc comments for classes and methods

adamzzq committed:
Update GitHub repo README and User Guide

adamzzq committed:
Update download link in README.md

adamzzq committed:
Add _config.yml file for customized UG theme

adamzzq committed:
Update _config.yml

adamzzq committed:
Update show_downloads configuration in _config.yml

adamzzq committed:
Add style.scss to widen the UG webpage layout

adamzzq committed:
Update User Guide webpage

adamzzq committed:
Fix bug: invalid find and add command
* User cannot add or find empty tasks
* Add more flexibility by allowing case-insensitive commands

adamzzq committed:
Update UG webpage

adamzzq committed:
Update file loading error message

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Update README.md

adamzzq committed:
Add exception handling for event time validation

adamzzq committed:
Add exception handling for event time validation

adamzzq committed:
Add exception handling for event time validation

adamzzq committed:
Merge branch 'master' into branch-time-error-handling

adamzzq committed:
Merge pull request #6 from adamzzq/branch-time-error-handling

Add exception handling for event time validation

adamzzq committed:
Merge branch 'master' of github.com:adamzzq/ip

# Conflicts:
#	src/main/java/ui/Time.java

adamzzq committed:
Refactor exit flag variable name in Adam class

adamzzq committed:
Update UG format

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update style.scss

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Update UG webpage layout

adamzzq committed:
Change banner width

adamzzq committed:
Set favicon.ico

adamzzq committed:
Set favicon.ico

adamzzq committed:
Configure head-custom.html

adamzzq committed:
Configure default.html

adamzzq committed:
Configure default.html

adamzzq committed:
Configure default.html

adamzzq committed:
Configure default.html

adamzzq committed:
Configure default.html

adamzzq committed:
Configure default.html

adamzzq committed:
Update UG page

adamzzq committed:
Fix minor details in JavaDoc

adamzzq committed:
Update UG webpage

adamzzq committed:
Update default.html

adamzzq committed:
Update default.html

adamzzq committed:
Update default.html
